crates/typst-pdf/src/image.rs

```diff
@@ -45,54 +46,56 @@ pub fn write_images(
                         space.icc_based(id);
                         icc_ref = Some(id);
-                    } else if *has_color {
-                        color::write(
-                            ColorSpace::Srgb,
-                            space,
-                            &context.globals.color_functions,
-                        );
                     } else {
                         color::write(
-                            ColorSpace::D65Gray,
+                            *color_space,
                             space,
                             &context.globals.color_functions,
                         );
                     }
 
                     // Add a second gray-scale image containing the alpha values if
                     // this image has an alpha channel.
-                    if let Some((alpha_data, alpha_filter)) = alpha {
+                    if let Some(alpha) = alpha {
                         let mask_ref = chunk.alloc.bump();
                         image.s_mask(mask_ref);
                         image.finish();
 
-                        let mut mask = chunk.image_xobject(mask_ref, alpha_data);
-                        mask.filter(*alpha_filter);
+                        let mut mask = chunk.image_xobject(mask_ref, &alpha.data);
+                        mask.filter(alpha.filter);
                         mask.width(*width as i32);
                         mask.height(*height as i32);
                         mask.color_space().device_gray();
-                        mask.bits_per_component(8);
+                        mask.bits_per_component(i32::from(alpha.bits_per_component));
+                        mask.interpolate(*interpolate);
                     } else {
                         image.finish();
                     }
 
-                    if let (Some(icc), Some(icc_ref)) = (icc, icc_ref) {
-                        let mut stream = chunk.icc_profile(icc_ref, icc);
+                    if let (Some(compressed_icc), Some(icc_ref)) =
+                        (compressed_icc, icc_ref)
+                    {
+                        let mut stream = chunk.icc_profile(icc_ref, compressed_icc);
                         stream.filter(Filter::FlateDecode);
-                        if *has_color {
-                            stream.n(3);
-                            stream.alternate().srgb();
-                        } else {
-                            stream.n(1);
-                            stream.alternate().d65_gray();
+                        match color_space {
+                            ColorSpace::Srgb => {
+                                stream.n(3);
+                                stream.alternate().srgb();
+                            }
+                            ColorSpace::D65Gray => {
+                                stream.n(1);
+                                stream.alternate().d65_gray();
+                            }
+                            _ => unimplemented!(),
```

Comment:
Here, we only ever write `\N 1` or `\N 3`, but not `\N 4` even though the original file could be CMYK.